tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/standard/ReTLALanguagePred.scala

```diff
@@ -0,0 +1,179 @@
+      // [f EXCEPT ![a,b,c,...] = r] (not [f EXCEPT ![a][b] = r] or [f EXCEPT ![a] = r, ![b] = s] )
+      case OperEx(TlaFunOper.except, fn, key, value) =>
+        isOkInContext(letDefs, fn)
+          .and(
+              isOkInContext(letDefs, value),
+          )
+          .and(
+              key match {
+                // ![a,b,...] case
+                case OperEx(TlaFunOper.tuple, OperEx(TlaFunOper.tuple, args @ _*)) =>
+                  allArgsOk(letDefs, args)
+                // ![a] case
+                case OperEx(TlaFunOper.tuple, args @ _*) =>
+                  allArgsOk(letDefs, args)
+                // Impossible, but we need case-completeness
+                case _ => PredResultFail(List())
+              },
+          )
+
+      case LetInEx(body, defs @ _*) =>
+        // go inside the let definitions (similar to FlatLanguagePred)
+        def eachDefRec(ctx: Set[String], ds: List[TlaOperDecl]): PredResult = {
+          ds match {
+            case Nil =>
+              PredResultOk()
+
+            case head :: tail =>
+              isOkInContext(ctx, head.body) // check the first operator definition
+                .and(eachDefRec(ctx + head.name, tail)) // check the other operator definitions
+          }
+        }
+
+        // check the let-definitions first, in a sequence, as they may refer to each other
+        val defsResult = eachDefRec(letDefs, defs.toList)
+        val newLetDefs = defs.map(_.name).toSet
+        // check the terminal expression in the LET-IN chain, by assuming the context generated by the definitions
+        defsResult
+          .and(isOkInContext(letDefs ++ newLetDefs, body))
+
+      case e @ OperEx(TlaOper.apply, NameEx(opName), args @ _*) =>
+        // the only allowed case is calling a nullary operator that was declared with let-in
+        if (!letDefs.contains(opName)) {
+          PredResultFail(List((e.ID, s"undeclared operator $opName")))
+        } else if (args.nonEmpty) {
+          PredResultFail(List((e.ID, s"non-nullary operator $opName")))
+        } else {
+          PredResultOk()
+        }
+
+      case e =>
+        PredResultFail(List((e.ID, e.toString)))
+    }
+  }
+}
+
+object ReTLALanguagePred {
+  private val singleton = new ReTLALanguagePred
+
+  protected val unaryOps: HashSet[TlaOper] =
+    HashSet(
+        TlaActionOper.prime,
+        TlaBoolOper.not,
+//        TlaArithOper.uminus,
+        ApalacheOper.skolem,
+    )
+
+  protected val binaryOps: HashSet[TlaOper] =
+    HashSet(
+        TlaOper.eq,
+        TlaOper.ne,
+        TlaBoolOper.implies,
+        TlaBoolOper.equiv,
+        // IntArith not in v1
+//        TlaArithOper.plus,
+//        TlaArithOper.minus,
+//        TlaArithOper.mult,
+//        TlaArithOper.div,
+//        TlaArithOper.mod,
+//        TlaArithOper.exp,
+//        TlaArithOper.lt,
+//        TlaArithOper.gt,
+//        TlaArithOper.le,
+//        TlaArithOper.ge,
```

Comment:
What's up with these commented out parts?

Reply:
There was a discussion on whether the 1st version of reTLA should include support for integer arithmetic or not. It's no problem in SMT, but not as great for IVy. Since we'll very likely be adding these operators in the future (and since my 1st draft had them), I just commented them out and made a note, so we can go to v2 just by uncommenting.

Reply:
Oh I see. I didn't understand what `IntArith not in v1` means without this context.